pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -506,6 +507,10 @@ func (m *kubeGenericRuntimeManager) getPodContainerStatuses(uid kubetypes.UID, n
 	// TODO: optimization: set maximum number of containers per container name to examine.
 	for i, c := range containers {
 		status, err := m.runtimeService.ContainerStatus(c.Id)
+		if crierror.IsNotFound(err) {
```

Comment:
We need tests to verify this behavior - one unit test for the generic runtime, and then probably an integration test.  We also want to add a node e2e test to verify that the behavior doesn't regress.  We're changing a fundamental property of the kubelet (one that was racy before), so we want to make sure in the future we get the right outcome, and can't regress.

Reply:
I would add a comment here explaining that between List (above) and check (here) another thread might remove a container, and that is normal.

Reply:
Done

Reply:
@smarterclayton I can mock runtimeservice to throw a notfound error in unit test, but I could not catch any thing that can illustrate that syncTerminatingFn has encountered  notfound error in e2e test by kube-api(no event or no pod status change), so do you thank we can write such a e2e test for regression?

Reply:
It should be possible to write a node e2e that creates the pod, then triggers a delete AND deletes the container.  However, for the sake of the fix, I don't think it's necessary here.